site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -2331,6 +2355,165 @@ public function ajaxSaveGradedComponent($gradeable_id) {
+        $component = $this->tryGetComponent($gradeable, $component_id);
+        if ($component === false) {
+            return;
+        }
+
+        // Get user id from the anon id
+        $submitter_id = $this->tryGetSubmitterIdFromAnonId($anon_id, $gradeable_id);
+        if ($submitter_id === false) {
+            return;
+        }
+
+        // Get the graded gradeable
+        $graded_gradeable = $this->tryGetGradedGradeable($gradeable, $submitter_id);
+        if ($graded_gradeable === false) {
+            return;
+        }
+
+
+        // checks if user has permission
+        if (!$this->core->getAccess()->canI("grading.electronic.save_graded_component", ["gradeable" => $gradeable, "graded_gradeable" => $graded_gradeable, "component" => $component])) {
+            $this->core->getOutput()->renderJsonFail('Insufficient permissions to save component/marks');
+            return;
+        }
+
+        $logger_params = [
+            "course_semester" => $this->core->getConfig()->getTerm(),
+            "course_name" => $this->core->getDisplayedCourseName(),
+            "gradeable_id" => $gradeable_id,
+            "grader_id" => $this->core->getUser()->getId(),
+            "component_id" => $component_id,
+            "action" => "OPEN_COMPONENT",
+            "submitter_id" => $submitter_id
+        ];
+        Logger::logTAGrading($logger_params);
+
+        $graders = $graded_gradeable->getActiveGraders();
+        $timestamps = $graded_gradeable->getActiveGradersTimestamps();
+        $graders_names = $graded_gradeable->getActiveGradersNames();
+        $this->core->getQueries()->addComponentGrader($component, $gradeable->isTeamAssignment(), $grader->getId(), $submitter_id);
+
+        if ($gradeable->hasPeerComponent() && !$this->core->getUser()->accessGrading()) {
+            // return empty data for peers
+            return JsonResponse::getSuccessResponse(['active_graders' => [], 'active_graders_timestamps' => []]);
+        }
+        // If there are no graders for this component, use an empty array
+        if (isset($graders[$component_id])) {
+            // Ensure the current grader is not in the list of active graders
+            for ($i = 0; $i < count($graders[$component_id]); $i++) {
+                if ($graders[$component_id][$i] === $grader->getId()) {
+                    array_splice($timestamps[$component_id], $i, 1);
+                    array_splice($graders_names[$component_id], $i, 1);
+                    break;
+                }
+            }
+        }
+
+        return JsonResponse::getSuccessResponse(['active_graders' => $graders_names, 'active_graders_timestamps' => $timestamps]);
+    }
+
+    /**
+     * @param string $gradeable_id
+     * @return JsonResponse|void
+     */
+    #[Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/graded_gradeable/close_component", methods: ["POST"])]
+    public function ajaxCloseGradedComponent($gradeable_id) {
```

Comment:
There's a lot of duplicated code between this method and `ajaxOpenGradedComponent()`.  Any way you could share some code here?